contracts/0.8.9/lib/WithdrawalRequests.sol

```diff
@@ -0,0 +1,122 @@
+// SPDX-FileCopyrightText: 2023 Lido <info@lido.fi>
+// SPDX-License-Identifier: GPL-3.0
+
+pragma solidity 0.8.9;
+
+library WithdrawalRequests {
+    address constant WITHDRAWAL_REQUEST = 0x0c15F14308530b7CDB8460094BbB9cC28b9AaaAA;
+
+    error MismatchedArrayLengths(uint256 keysCount, uint256 amountsCount);
+    error FeeNotEnough(uint256 minFeePerRequest, uint256 requestCount, uint256 msgValue);
+
+    error WithdrawalRequestFeeReadFailed();
+    error InvalidPubkeyLength(bytes pubkey);
+    error WithdrawalRequestAdditionFailed(bytes pubkey, uint256 amount);
+    error NoWithdrawalRequests();
+    error PartialWithdrawalRequired(bytes pubkey);
+
+    event WithdrawalRequestAdded(bytes pubkey, uint256 amount);
+
+    /**
+     * @dev Adds full withdrawal requests for the provided public keys.
+     *      The validator will fully withdraw and exit its duties as a validator.
+     * @param pubkeys An array of public keys for the validators requesting full withdrawals.
+     */
+    function addFullWithdrawalRequests(
+        bytes[] calldata pubkeys
+    ) internal {
+        uint256 keysCount = pubkeys.length;
+        uint64[] memory amounts = new uint64[](keysCount);
+
+        _addWithdrawalRequests(pubkeys, amounts);
+    }
+
+    /**
+     * @dev Adds partial withdrawal requests for the provided public keys with corresponding amounts.
+     *      A partial withdrawal is any withdrawal where the amount is greater than zero.
+     *      This allows withdrawal of any balance exceeding 32 ETH (e.g., if a validator has 35 ETH, up to 3 ETH can be withdrawn).
+     *      However, the protocol enforces a minimum balance of 32 ETH per validator, even if a higher amount is requested.
+     * @param pubkeys An array of public keys for the validators requesting withdrawals.
+     * @param amounts An array of corresponding withdrawal amounts for each public key.
+     */
+    function addPartialWithdrawalRequests(
```

Comment:
What if I want a batch that is mix of partial and full withdrawals?

Reply:
The separation of Full and Partial withdrawals was suggested to simplify an interaction with the withdrawal contract.

An approach, proposed in the [Triggerable Withdravals](https://github.com/lidofinance/core/pull/9040) document suggests that the Validator Exit Bus will only use full withdrawal. Usages of partial withdrawals might require a different mechanism inside the Lido protocol, which is why this PR proposes to separate those types of requests.

If you consider that a mix of partial and full withdrawals might be useful in the future we can consider multiple approaches. 

1. Add the `addWithdravalCredentials` method which will consume both partial and full withdrawals.

2. In addition to the first approach, remove `addPartialWithdrawalRequests` method, so the `addWithdravalCredentials` will be used for both partials and mix (partial + full) withdrawal requests.

What do you think will be an optimal approach?

Reply:
1st one will be fine. I think that this kind of constraint is not required in vaults.